Cookbook/Proposals/ModelFixtures.md

```diff
@@ -0,0 +1,249 @@
+> The term "mock" has a precise definition as well, as a specific sub-type of test double. A mock object asserts that certain invocations are made on itself, and will raise exception as soon as any unexpected interactions take place. [[source](https://github.com/testdouble/contributing-tests/wiki/Mock#precise-use)]
+
+- Naming of namespacing enums is inconsistent e.g. `MockPatientDetailsDTO` for `PatientDTO`.
+- Sometimes there are multiple `make` helpers to maintain. See more details about issues with this in "Object Mother" section in "Altenratives considered".
+
+In our codebase there are multiple implementations of this pattern with variations in naming and functionality, e.g.
+
+- `MessageTestData.makeMessages`,
+- `MessageTestData.makeMessageResponse`,
+- `MockPatientDetailsDTO.makeWithPartner`,
+- `MockAddressDTO.make`,
+- `CoreSessionDTO.mock`.
+
+On top of that some tests create value type by directly calling the initializer.
+
+Those inconsistencies cause friction when trying to setup state in tests. It's hard to discover existing test helpers. Usage of unparametrized factory methods (e.g. `MockAddressDTO.make()`) makes it hard to understand on what data the test depends.
+
+## Proposed solution
+
+<!-- // This section should the answer the `how?`. -->
+<!-- Describe your solution to the problem. Provide examples and describe
+how they work. Show how your solution is better than current
+workarounds: is it cleaner, safer, or more efficient? -->
+
+Proposed approach is similar to what we often do by the convention with `MockSomeNameDTO.make` pattern. It aims to improve current situation by introducing a few important rules.
+
+1. Helpers for models should be defined as an **extension of the model type** in a test target or framework with test utilities.
+1. Helper method should be named `fixture` to signify that it's part of the tests.
+1. Helpers should use parameters that mirror model's initializer.
+1. All of helper method parameters should have default value. This will make relevant data stand out and irrelevant data will be provided by defaults.
+
+Let's see sample implementation. First define extension on an existing type, instead of introducing extra type for namespacing.
+
+```swift
+// Example for points 1. and .2
+
+struct Foo {
+    let a: Int
+    let b: String
+}
+
+// add extension in test target or test utilities framework:
+
+extension Foo {
+    static func fixture() -> Foo {
+        return Foo(a: 1, b: "xyz")
+    }
+}
+```
+
+Building upon this, use the same arguments in `fixture` factory method as in the models's initializer. Provide a default value for each argument. Those values should be "reasonable defaults", e.g. for identifier it can be `"1"` value. Example of adjusted `Foo.fixture` method:
+
+```swift
+// Example for points 3. and .4
+
+extension Foo {
+    static func fixture(a: Int = 1, x: String = "xyz") -> Foo {
```

Comment:
👍 
just btw. I used to call it `static func fake(...)`.
Just to consider if we like the name more.
However, I'm fine with `fixture`

Reply:
As much as I love the approach and proposal, am not a huge fan of a fancy name `fixture`.  Could we make it something more obvious like `Foo.testObject`?

Reply:
(Sorry for unsolicited review! Don't know what the system is around that) 😬

Reply:
@TheAdamBorek  @chitrakotwani I'd be personally fine with either `fake` or `testObject`, but I guess it's a matter of habit. Strictly speaking though it is not necessary a `fake` double, because it can be actually used by tested code, if we use Fouler's blog post as a reference for classification. And fixtures he mentioned [here](https://www.martinfowler.com/articles/mocksArentStubs.html#FixtureSetup) but he does not mention this as a type of test doubles (he does mention Object Mother pattern). It's something that I used to see in other languages like Ruby and JavaScript so seems like a common term for this kind of stuff. I'd even suggest to use the same name instead of our `canned responses`, which is another weird name in fact.

Reply:
Regarding the naming of method issue. Options I see:

- keep `fixture`,
- add `test` prefix, i.e. `testFixture`.

I don't see that anybody is strongly against keeping the original name. It should also be simple enough to rename it afterwards if we find it problematic after adoption.

> It's something that I used to see in other languages like Ruby and JavaScript so seems like a common term for this kind of stuff.

I used the name after Ilya's suggestion as it nicely expresses the purpose of helper. Note about "fixture" term from xUnit Patterns:

> In generic xUnit a test fixture is all the things we need to have in place in order to run a test and expect a particular outcome.

More details:
- [test fixture (in xUnit)](http://xunitpatterns.com/test%20fixture%20-%20xUnit.html),
- [test fixture (disambiguation)](http://xunitpatterns.com/test%20fixture%20-%20ambiguous.html).

---

> just btw. I used to call it static func fake(...) just because I used to use Faker gem in my ruby tests 😅

@TheAdamBorek Most of the time those are value types, so they're real values, only configured for tests. I think `fake` would give wrong impression.

> As much as I love the approach and proposal, am not a huge fan of a fancy name fixture. Could we make it something more obvious like Foo.testObject?

The `Object` could be misleading, because most of the times our models are value types.

> (Sorry for unsolicited review! Don't know what the system is around that) 😬

@chitrakotwani  All feedback is welcome! Thanks for reviewing. Ideally everyone should review. The more, the better! For example I might be too used to this approach, so it helps me with understanding other perspectives and see what needs documenting.